An Introduction to Unit Testing

fabioscagliola opened the pull request:
This article provides an introduction to unit testing.

fabioscagliola commented:
Apologies, guys: for some reason I cannot select any reviewer. I am pretty sure I did something wrong. So, I am tagging you here. Please, bear with me. @mohsenr @r3dDoX @lehmamic @linkdotnet

linkdotnet commented:
Hey @fabioscagliola, welcome welcome.

I am currently a bit busy, so I could only have a brief look into your post.
For starters a really cool and detailed guide about unit testing. 

You used `c#` as fence for csharp code. To be honest, I can not tell you whether or not hashnode supports that fence. To be sure, we could switch to `csharp` instead. Alternatively we just publish and check if code highlighting is working.

Also the formatting seems a bit odd. Every second line is a new line, which makes reading it a bit harder. I don't know if this is related to line-endings and how GitHub displays them.

I will report back content wise.

r3dDoX commented:
Hey @fabioscagliola, good point I will check the repo settings but added me as a reviewer for now. Will check as soon as possible. What I noticed right away in here is that your file might miss the .md ending.

r3dDoX commented:
Please have a look at the README.md in this repo, there is some meta data you have to put to the top of your markdown file for it to be properly published on Hashnode

fabioscagliola commented:
Thank you, @linkdotnet and @r3dDoX 

I have just implemented the following changes:

- I added the MD extension to the file 
- I added the header 
- I changed the fenced code blocks from `c#` to `csharp` 
- I removed the blank lines

r3dDoX commented:
@fabioscagliola thanks for the changes, looking good! For us to be able to publish this blog eventually we will need you as a user to join our team blog on Hashnode. Please follow this invite: https://hashnode.com/teams/invite/a27b3c7d-d4c1-4dad-ad81-a9fab11e301f and make sure you're in our team?

I saw you don't have a cover image. If you want to, you can either upload a picture on hashnode and reference it in the cover meta. A quick guide on this you can find in README.md in this repo. Otherwise, I do often use a picture from unsplash.

fabioscagliola commented:
@r3dDoX, I have just created an account on Hashnode and accepted you invitation, and I have added a cover. Thank you

linkdotnet commented:
Hey @fabioscagliola I did read the article and good job on this one.
Personally one request (highly opinionated): Can we condense some of the code samples or content in general?

fabioscagliola commented:
Hello @linkdotnet, I revised the code samples: I realize that there is quite some duplicated code, especially in the last ones, as they are very similar. However, for the sake of completeness of each sample, I would rather not remove this duplicated code. Other than that, the textual part is pretty short, and I would rather not delete anything. If this is not a showstopper, and you really value the contents, I would kindly ask you to publish the article as is. Thank you in advance. Fabio

linkdotnet commented:
Hey @fabioscagliola, no showstopper at all and I am on board with the blog post.
Unfortunately, I don't have the rights to merge or publish it :D
**But** I can give my symbolic approval:D

lehmamic commented on an-introduction-to-unit-testing.md:
```diff
@@ -0,0 +1,852 @@
+---
```
Could you move the file into the drafts folder, please?

lehmamic commented on an-introduction-to-unit-testing.md:
```diff
@@ -0,0 +1,852 @@
+---
+title: An Introduction to Unit Testing
+domain: campzulu.hashnode.dev
+tags: Unit Testing, C#, NUnit
```
You need to use the slugs of the tags: https://github.com/Hashnode/support/blob/main/misc/tags.json

Your tags don't exist in that list. I think hashnode will not accept that.

Possible fit:
```ms
tags: unit-testing, csharp
```

lehmamic replied:
maybe `testing` as ell

lehmamic commented on an-introduction-to-unit-testing.md:
```diff
@@ -0,0 +1,852 @@
+
+## Unit testing in theory
+
+This section defines unit testing, units of work, integration tests, and test-driven development.
```
Maybe you also could loose a word about the testing pyramid?

lehmamic commented on an-introduction-to-unit-testing.md:
```diff
@@ -0,0 +1,852 @@
+
+### Integration tests
+
+Integration tests are any tests that aren’t fast and consistent, and that use one or more real dependencies of the units under test. If the test uses the real system time, the real filesystem, a real database, or some piece of hardware, it has stepped into the realm of integration testing.
```
In my point of view a test always needs to be consistent. Consistent means complete reproducable without any "flickery" aspect.

lehmamic replied:
Maybe also an advantage is that it tests the classes in a combination which doesn't happen in unit tests. Also with side effects which are usually mocked away in unit tests.

lehmamic commented on an-introduction-to-unit-testing.md:
```diff
@@ -0,0 +1,852 @@
+
+### Integration tests drawbacks
+
+If a test uses the real database, then it’s no longer only running in memory; in that, its actions are harder to erase than when using only in-memory fake data. Integration tests increase the risk of another problem: testing too many things at once. In an integration test you can have many failure points and, therefore, finding the cause of the problem is harder. An integration test uses real dep ...
```
What is the recipe for this? Integration tests need to be executed in a well-defined, controlled sandbox.

lehmamic commented on an-introduction-to-unit-testing.md:
```diff
@@ -0,0 +1,852 @@
+
+## Unit testing in practice
+
+This section introduces the xUnit testing frameworks.
```
here you are writing xunit, in the tags you write nunit....

lehmamic commented on an-introduction-to-unit-testing.md:
```diff
@@ -0,0 +1,852 @@
+
+### The xUnit frameworks
+
+xUnit is a collective name for several unit testing frameworks that derive their structure and functionality from Smalltalk’s SUnit, designed by Kent Beck in 1998.
```
I would directly add some links to lead to your source

lehmamic commented on an-introduction-to-unit-testing.md:
```diff
@@ -0,0 +1,852 @@
+
+The name of a unit test usually includes the name of the class, the name of the unit of work, and the expected behavior.
+
+Example:  `OpticalSensor_IsRedBloodCell_ReturnsFalse()`
```
There are many naming conventions, some projects for example are using a "given..._when..._then..." format

lehmamic commented on an-introduction-to-unit-testing.md:
````diff
@@ -0,0 +1,852 @@
+
+The **UnitTesting01.cs** file includes two unit tests that verify the Boolean values returned by the  `IsRedBloodCell`  method of the  `OpticalSensor`  class.
+
+```csharp
````
Just a suggestion: Write the file name with the inline code style as well. e.g. `UnitTest01.cs`

lehmamic commented on an-introduction-to-unit-testing.md:
````diff
@@ -0,0 +1,852 @@
+}
+```
+
+## Suggested readings
````
I suggest that you add links to that sources.

lehmamic commented on an-introduction-to-unit-testing.md:
```diff
@@ -0,0 +1,852 @@
+
+### Fakes: stubs and mocks
+
+A fake is a generic term that can describe either a stub or a mock. A stub is a controllable replacement for an existing dependency: by using a stub, you can test your code without dealing with the dependency directly. Mocks are just like stubs, but you can assert against the mock object, whereas you do not assert against a stub.
```
I'm not sure, but I think mocking framework such as Nsubstitute or Moq are also mentionable and maybe as an extension to that some AutoMock addon to those frameworks. Because it reduces a lot of maintenance effort in your tests. But that's up to you.

lehmamic reviewed (changes requested):
Its a good article. I added a lot of comments and suggestions. Its up to you what you want to take over.

One thing you need to fix:
The tags are not conform to the hashnode tag list. See my comment there.

fabioscagliola commented:
Hello @lehmamic, thank you for your comments and suggestions, which I appreciate and agree with. Unfortunately, at the moment, I am too busy to revise the article, therefore I have to stick to the bare necessities: I revised the tags and moved the file to the **drafts** folder. If this is not a showstopper, and you really value the contents, I would kindly ask you to publish the article as is. Thank you once more. Fabio
